### ISSUE TITLE
sql: support `record` types as schemalass `ScalarType::Record`

### ISSUE THREAD

**Issue by sploiselle** [MaterializeInc/materialize#11580]
### Feature request

tl;dr: anonymous records are currently impossible to support

In PG, the `record` type represents the anonymous row type; you know it's a row, but its schema can be anything--arbitrary columns of ints, dates, other records, etc.

Any operation that relies on "type equality" (e.g. union, homogenizing input) can take `record` types as their input. When the input type is `record`, so too should the output type be `record`. For example:

```
SELECT least(row(1), row(4, 5));
(1)

SELECT pg_typeof(least(row(1), row(4, 5)));
 record
```
```
SELECT least(row(1,5), row(4));
(1,5)

SELECT pg_typeof(least(row(1,5), row(4)));
 record
```

```sql
SELECT * FROM (VALUES (1, (1,))) AS l
UNION SELECT * FROM (VALUES (2, (2, 3))) AS r;
```
```
 column1 | column2
---------+---------
       1 | (1)
       2 | (2,3)
```

```sql
SELECT
	DISTINCT pg_typeof(column2)
FROM
	(
		SELECT * FROM (VALUES (1, (1))) AS l
		UNION SELECT * FROM (VALUES (2, (2, 3))) AS r
	)
		AS foo;
```
```
record
```

However, Materialize currently has no means by which to permit these kind of arbitrarily-schema'ed rows. The easiest to encounter place to trigger this inability is in `pgrepr::Value::from_datum`, which takes a `Datum` and a `ScalarType` into a `pgrepr::Value`.

`ScalarType::Record` uses a `Datum::List` as its backing and then re-interprets the elements of the list as the appropriate type based on the `fields` in `ScalarType::Record`. However, in the case of the PG-like `record` type, we do not have a meaningful `fields` type, which means we don't know how to encode the `record` into a value.

We cannot naively/blithely convert the `DatumList`'s `Datum`s into the correct `Value`s because multiple `ScalarType`s are backed by the same `Datum`s, i.e. we would lose the ability to differentiate between `String` and `Char`, which is meaningful because we must resaturate `Char` with the appropriate number of spaces when sending them to PG.

We also cannot more forcefully infer the schema of the return rows. As you can see in e.g. `least`, they're evaluated at runtime.

# Potential solutions

- Permit `ScalarType::Record` to represent anonymous `record` types, but insist that places that use them in multiples (i.e. homogenous operations, union, etc.) only use a single schema. This, in essence, makes `record` into a polymorphic type that ensures that all instance of it share a common type.

    This is how the code works today, but prohibits us from supporting differently-schema'ed `record`s in e.g. homogenizing functions.

- Include the `ScalarType`-ed schema in `DatumList` by means of expanding `repr::row::Tag`. I believe this is a problem for storage, which wants to use as few of these as possible. There's also a question about how to represent custom types in this schema.

   However, PG [does something akin to this](https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/rowtypes.c#L929), so it seems unlikely we'll be able to reach feature parity without it.

- Collect a bag of possible schemas based on the input--those are the candidates schemas that can get returned. When returning a value, select the best of them based on the `DatumList` schema; this becomes akin to function selection. I think this solution is fraught but the thought occurred to me, so jotting it down.
